bundles/org.openhab.binding.teleinfo/src/main/java/org/openhab/binding/teleinfo/internal/serial/TeleinfoSerialControllerHandler.java

```diff
@@ -0,0 +1,196 @@
+    private long invalidFrameCounter = 0;
+
+    public TeleinfoSerialControllerHandler(Bridge thing, SerialPortManager serialPortManager) {
+        super(thing);
+        this.serialPortManager = serialPortManager;
+    }
+
+    @Override
+    public void initialize() {
+        invalidFrameCounter = 0;
+
+        keepAliveThread = scheduler.scheduleWithFixedDelay(() -> {
+            if (!isInitialized()) {
+                openSerialPortAndStartReceiving();
+                updateStatus(ThingStatus.UNKNOWN);
+            }
+            logger.debug("Check Teleinfo receiveThread status...");
+            logger.debug("isInitialized() = {}", isInitialized());
+            TeleinfoReceiveThread receiveThreadRef = receiveThread;
+            if (receiveThreadRef != null) {
+                logger.debug("receiveThread.isAlive() = {}", receiveThreadRef.isAlive());
+            }
+            if (isInitialized() && (receiveThreadRef == null || !receiveThreadRef.isAlive())) {
+                updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.NONE, ERROR_UNKNOWN_RETRY_IN_PROGRESS);
+                logger.info("Try to restart Teleinfo receiving...");
+                stopReceivingAndCloseSerialPort();
+                openSerialPortAndStartReceiving();
+            }
+        }, 0, 60, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public void dispose() {
+        ScheduledFuture<?> keepAliveThreadRef = keepAliveThread;
+        if (keepAliveThreadRef != null) {
+            keepAliveThreadRef.cancel(true);
+            keepAliveThread = null;
+        }
+        stopReceivingAndCloseSerialPort();
+
+        super.dispose();
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+    }
+
+    @Override
+    public void onFrameReceived(TeleinfoReceiveThread receiveThread, Frame frame) {
+        updateStatus(ThingStatus.ONLINE);
+        fireOnFrameReceivedEvent(frame);
+    }
+
+    @Override
+    public void onInvalidFrameReceived(TeleinfoReceiveThread receiveThread, InvalidFrameException error) {
+        invalidFrameCounter++;
+        updateState(THING_SERIAL_CONTROLLER_CHANNEL_INVALID_FRAME_COUNTER, new DecimalType(invalidFrameCounter));
+    }
+
+    @Override
+    public void onSerialPortInputStreamIOException(TeleinfoReceiveThread receiveThread, IOException e) {
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, ERROR_UNKNOWN_RETRY_IN_PROGRESS);
+    }
+
+    @Override
+    public boolean continueOnReadNextFrameTimeoutException(TeleinfoReceiveThread receiveThread, TimeoutException e) {
+        logger.warn("Retry in progress. Next retry in {} seconds...", SERIAL_PORT_DELAY_RETRY_IN_SECONDS);
+        updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.NONE, ERROR_UNKNOWN_RETRY_IN_PROGRESS);
+        try {
+            Thread.sleep(SERIAL_PORT_DELAY_RETRY_IN_SECONDS * 1000);
+            return true;
+        } catch (InterruptedException e1) {
+            return false;
+        }
```

Comment:
I would move this sleep to the reader thread since a sleep should never be done in a scheduled task on the openHAB `scheduler`. To make sure it's clear this sleep is not associated with this class.